keystone/common/api_filtering.py

```diff
@@ -0,0 +1,210 @@
+                if res['user'].get('default_project_id') is not None:
+                    if (res['user']['default_project_id']
+                            in project_details.values()):
+                        msg = ('User \'%s\' does not exist.'
+                               % kwargs.get('user_id'))
+                        LOG.error(_LE(msg))
+                        raise exception.NotFound(msg)
+            else:
+                refs = [i for i in res['users']]
+
+                for i in refs:
+                    if (i.get('default_project_id') is not None and
+                            i['default_project_id']
+                            in project_details.values()):
+                        res['users'].remove(i)
+
+                    if (i.get('name')
+                            in CONF.api_filtering.explicit_users_to_filter):
+                        res['users'].remove(i)
+        return res
+    return wrapper
+
+
+def get_roles(self, context):
+    try:
+        user_token_ref = token_model.KeystoneToken(
+            token_id=context['token_id'],
+            token_data=self.token_provider_api.validate_token(
+                context['token_id']))
+    except exception.TokenNotFound as e:
+        LOG.error(_LE('Token not found'))
+        raise exception.Unauthorized(e)
+
+    creds = copy.deepcopy(user_token_ref.metadata)
+
+    try:
+        creds['user_id'] = user_token_ref.user_id
+    except exception.UnexpectedError:
+        LOG.error(_LE('Invalid user'))
+        raise exception.Unauthorized()
+
+    LOG.debug(_LE("user_token_ref project_scoped = %s "
+                  % user_token_ref.project_scoped))
+
+    if user_token_ref.project_scoped:
+        creds['tenant_id'] = user_token_ref.project_id
+    else:
+        LOG.error(_LE('Invalid tenant / project not scoped'))
+
+    creds['roles'] = user_token_ref.role_names
+
+    return creds['roles']
+
+
+def is_admin(self, context):
+
+        """
+        It indicates whether the performing actor is 'admin' or not.
+        """
+        if not context['is_admin']:
+            if 'admin' in get_roles(self, context):
+                return True
+            else:
+                return False
+        else:
+            return True
+
+
+def v2_block_user(f):
+
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+
+        res = f(*args, **kwargs)
+        is_admin_flag = is_admin(args[0], args[1])
+
+        if not is_admin_flag:
+            flag = False
+            roles = get_roles(args[0], args[1])
+            for role in roles:
+                if role in CONF.api_filtering.admin_roles:
+                    flag = True
+
+            if not flag:
+                err_msg = ("You are not authorized to perform this action: "
+                           "get_project_users")
+                LOG.error(_LE(msg))
+                raise exception.Unauthorized(err_msg)
+
+        return res
+    return wrapper
+
+
+def list_projects(self, context):
```

Comment:
`list_projects` appears to be a function name used in other bits of keystone. Since this version seems to be intended for use only within this module, it would be better to prepend a `_` to indicate that it's private.